File: bench/kernels/e1_sum_reduce/ref.c
/* e1-sum-reduce (family E), naive C — what people actually write.
 * i64 summation with statically boundable ranges; C's signed overflow is
 * UB, so clang assumes it away and vectorizes the reduction freely. That
 * is precisely the fact wolf renounced (X3: checked arithmetic in every
 * profile), so this kernel prices the renunciation.
 * Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static int64_t sum_reduce(int64_t n) {
    int64_t acc = 0;
    for (int64_t i = 0; i < n; i++) acc = acc + (i & 1023);
    return acc;
}

int main(int argc, char **argv) {
    int64_t ops = argc > 1 ? (int64_t)strtoull(argv[1], 0, 10) : 4000;
    const int64_t inner = 100000;
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    int64_t sink = 0;
    for (int64_t k = 0; k < ops; k++) sink = sink % 4096 + sum_reduce(inner) % 4096;
    clock_gettime(CLOCK_MONOTONIC, &t1);
    uint64_t ns = (uint64_t)(t1.tv_sec - t0.tv_sec) * 1000000000ull
                + (uint64_t)(t1.tv_nsec - t0.tv_nsec);
    printf("{\"ns\":%llu,\"ops\":%lld,\"sink\":%lld}\n",
           (unsigned long long)ns, (long long)(ops * inner), (long long)sink);
    return 0;
}

File: bench/kernels/e1_sum_reduce/ref.rs
//! e1-sum-reduce (family E), Rust reference. Release Rust checks nothing
//! (wrapping in debug is a debug_assert; `-O` uses plain `add`), so this
//! lane sits between naive C and wolf's always-checked arithmetic.
//! Protocol: argv[1]=ops; prints {"ns":..,"ops":..,"sink":..}.

use std::hint::black_box;

fn sum_reduce(n: i64) -> i64 {
    let mut acc: i64 = 0;
    let mut i: i64 = 0;
    while i < n {
        acc = acc + (i & 1023);
        i += 1;
    }
    acc
}

fn main() {
    let ops: i64 = std::env::args()
        .nth(1)
        .and_then(|s| s.parse().ok())
        .unwrap_or(4000);
    let inner: i64 = 100_000;
    let t0 = std::time::Instant::now();
    let mut sink: i64 = 0;
    for _ in 0..ops {
        sink = sink % 4096 + sum_reduce(black_box(inner)) % 4096;
    }
    let ns = t0.elapsed().as_nanos();
    println!(
        "{{\"ns\":{ns},\"ops\":{},\"sink\":{sink}}}",
        ops.wrapping_mul(inner)
    );
}
